src/main/java/io/confluent/connect/elasticsearch/OffsetTracker.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2018 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.connect.elasticsearch;
+
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.connect.sink.SinkRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static java.util.stream.Collectors.toMap;
+
+/**
+ * Tracks processed records to calculate safe offsets to commit.
+ * <br>
+ * Since ElasticsearchClient can potentially process multiple batches asynchronously for the same
+ * partition, if we don't want to wait for all in-flight batches at the end of the put call
+ * (or flush/preCommit) we need to keep track of what's the highest offset that is safe to commit.
+ * For now, we do that at the individual record level because batching is handled by BulkProcessor,
+ * and we don't have control over grouping/ordering.
+ */
+class OffsetTracker {
+
+  private static final Logger log = LoggerFactory.getLogger(OffsetTracker.class);
+
+  private final Map<TopicPartition, Map<Long, OffsetState>> offsetsByPartition = new HashMap<>();
+  private final Map<TopicPartition, Long> maxOffsetByPartition = new HashMap<>();
+
+  private final AtomicLong numEntries = new AtomicLong();
+
+  static class OffsetState {
+
+    private final long offset;
+    private volatile boolean processed;
+
+    OffsetState(long offset) {
+      this.offset = offset;
+    }
+
+    /**
+     * Marks the offset as processed (ready to report to preCommit)
+     */
+    public void markProcessed() {
+      processed = true;
+    }
+
+    public boolean isProcessed() {
+      return processed;
+    }
+  }
+
+  /**
+   * Partitions are no longer owned, we should release all related resources.
+   */
+  public synchronized void closePartitions(Collection<TopicPartition> topicPartitions) {
+    topicPartitions.forEach(tp -> {
+      Map<Long, OffsetState> offsets = offsetsByPartition.remove(tp);
+      if (offsets != null) {
+        numEntries.getAndAdd(offsets.size());
```

Comment:
Why increment `numEntries` by the number of items in the map for the removed partition?

Reply:
Good catch! We should be decrementing here. Added asserts to cover it.